Add new major release policy to doc

maxime-rainville opened the pull request:
This adds a new major release policy.

Parent issues
- https://github.com/silverstripe/silverstripe-framework/issues/10345
- https://github.com/silverstripeltd/product-issues/issues/552

maxime-rainville commented:
I've added a new "Governance" section for things that aren't strictly about writing code.

michalkleiner commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+- new minor releases.
+
+A major release line stays in the "bug and security fixes" phase for 1 year.
+### Security fixes only
```
```suggestion

### Security fixes only
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+This policy applies to all commercially supported modules.
+
+Community modules are not covered by this policy. Modules under the `silverstripe` vendor name that are not commercially supported are updated on a best effort basis.
```
```suggestion
Community modules are not covered by this policy. Modules in the `silverstripe` github organisation that are not commercially supported are updated on a best effort basis.
```
Anyone can arbitrarily call their module `silverstripe/my-module` on composer and therefore would be under the `silverstripe` vendor name.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+## Scope of the policy
+
+This policy applies to all commercially supported modules.
```
Link to the commercially supported module page

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Community modules are not covered by this policy. Modules under the `silverstripe` vendor name that are not commercially supported are updated on a best effort basis.
+
+## General approach to Major releases
```
```suggestion
## General approach to major releases
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+## General approach to Major releases
+
+Silverstripe CMS aims to deliver regular major releases at predefined intervals with a clear support timeline. The key objectives of this policy is to allow Silverstripe CMS project owners to plan major upgrades ahead of time
```
```suggestion
Silverstripe CMS aims to deliver regular major releases at predefined intervals with a clear support timeline. The key objective of this policy is to allow Silverstripe CMS project owners to plan major upgrades ahead of time
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Silverstripe CMS aims to deliver regular major releases at predefined intervals with a clear support timeline. The key objectives of this policy is to allow Silverstripe CMS project owners to plan major upgrades ahead of time
+
+The main function of major releases is to ship *breaking changes* that can not be shipped in a minor releases. These changes are primarily aimed at making sure the major releases can be supported for the full duration of its lifecycle.
```
```suggestion
The main function of major releases is to ship *breaking changes* that can not be shipped in minor releases. These changes are primarily aimed at making sure the major releases can be supported for the full duration of its lifecycle.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+- pre-release (undefined duration)
+- active development (2 years)
+- bug and security fixes (1 year)
+- security fixes only (1 year).
```
```suggestion
- security fixes only (1 year)
```

maxime-rainville replied:
Having a dot on the final item in a list is the the recommended approach in our [writing style guide](https://silverstripe.atlassian.net/wiki/spaces/MAR/pages/506364107/Silverstripe+Writing+Style+Guide).

> The first type of bulleted list leads on from an introductory statement. All the items must match grammatically and complete the lead-in statement. Use a colon at the end of the lead-in statement. Put a full stop only at the end of the last bullet point.

ChloeSartorelli replied:
Interesting.... I don't personally like it but I can hardly argue with policy. I'll make a note to refer to this guide for the linting issue as well.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+- bug and security fixes (1 year)
+- security fixes only (1 year).
+
+The primary focus of major releases is to:
```
We've already talked about the "key objectives" and "main function", and now we're talking "primary focus" - while slightly different, these all have very similar meanings. Maybe these should be put together?

maxime-rainville replied:
I'll use something more neutral like "Most of the changes shipped in a major release will:"

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+The primary focus of major releases is to:
+- upgrade third party dependencies
+- remove and clean up deprecated APIs
```
How does "clean up" differ from "remove"?

maxime-rainville replied:
I was trying to communicate that we might change API that have not been deprecated but are confusing.

I'll try `remove deprecated APIs and clean up ambiguous APIs`

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+The primary focus of major releases is to:
+- upgrade third party dependencies
+- remove and clean up deprecated APIs
+- implement architectural changes that can not reasonably be introduced in a minor.
```
```suggestion
- implement architectural changes that can not reasonably be introduced in a minor release
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Major releases are not backward compatible. Most projects will require at least some development work to upgrade to a new major release.
+
+Silverstripe CMS provides a clear upgrade path to new major releases. This includes clear documentation on what breaking changes have been made to each area of Silverstripe CMS.
```
Maybe link to the "Upgrading" section of the docs

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+### Pre-release
+
+A Silverstripe CMS stable major release is preceded by beta period of at least three month. 
```
```suggestion
A Silverstripe CMS stable major release is preceded by a beta period of at least three months.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+- allow early adopters to plan their upgrade
+- identify bugs in the major release
+- help community module add support for the new major early on.
```
```suggestion
- help community modules add support for the new major release early on
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+### Stable release
+
+New major releases of Silverstripe CMS are tagged between April and June of odd years.
```
Weren't we talking about a March release? Where did April to June come from?

maxime-rainville replied:
I kept saying March when I meant to say first quarter of the year. April first is the start of our financial year.

Also to have any luck of releasing in March, we pretty much have to tag the beta before the end of the previous year.

If you think the March vs April makes a difference, I'm happy to stick to March.

ChloeSartorelli replied:
I'm perfectly happy with going with the first quarter of our financial year - just wanted to make sure I hadn't missed something there. The more time we have for this first one the better imo.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+### Active development
+
+Once a major release is stable, it enters a period of active development. A major release in active development:
+- Will receive regular minor releases that ships new feature and API in a backward compatible way.
+- Will receive regular patches for low impact bugs and low severity security issues.
```
```suggestion
- will receive regular minor releases that ships new feature and API in a backward compatible way
- will receive regular patches for low impact bugs and low severity security issues
```
Makes it consistent with the other lists formatting

maxime-rainville replied:
I'll re-gig those a bit to use the present tense and separate security issues from regular bugs.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+- Will receive regular minor releases that ships new feature and API in a backward compatible way.
+- Will receive regular patches for low impact bugs and low severity security issues.
+
+A major releases line will stay in "active development" of approximatively two years. "Active development" for a major release line ends when the follow up major release is tagged stable.
```
```suggestion
A major releases line will stay in "active development" for approximatively two years. "Active development" for a major release line ends when the next major release is tagged stable.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+A major releases line will stay in "active development" of approximatively two years. "Active development" for a major release line ends when the follow up major release is tagged stable.
+
+Only one major release line is in "active development" at once.
```
```suggestion
Only one major release line is in "active development" at a time.
```
or "at any given time"

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Release lines in the "bug and security fixes" phase will not receive:
+- new feature or new APIs
+- new minor releases.
```
```suggestion
- new minor releases
```

maxime-rainville replied:
View writing style guide.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+A Silverstripe CMS major release line enters the "bug and security fixes" phase once the follow major release is tagged stable.
+
+Release lines in the "bug and security fixes" phase will receive:
+- patches for low impact bugs
+- patches for low low severity security issues. 
```
```suggestion
- patches for low impact bugs
- patches for low severity security issues
```
Only low? Maybe something like "patches for bugs of any impact level" would be more appropriate?

maxime-rainville replied:
I've reworded those entries to align with the language used in the previous section.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+A major release line stays in the "bug and security fixes" phase for 1 year.
+### Security fixes only
+
+At the end of "Bug and security fixes" phase, a Silverstripe CMS major release line transition to the "security fixes only" phase.
```
```suggestion
At the end of "Bug and security fixes" phase, a Silverstripe CMS major release line transitions to the "security fixes only" phase.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+### End-of-life
+
+A Silverstripe CMS major release line will be end-of-life once its follow up major release line exit active development.
```
```suggestion
A Silverstripe CMS major release line will be end-of-life once the next major release line exits active development.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+A Silverstripe CMS major release line will be end-of-life once its follow up major release line exit active development.
+
+End-of-life major release do not receive any security fixes or bug fixes. 
```
```suggestion
End-of-life major releases do not receive updates of any kind including security fixes or bug fixes. 
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+## PHP Support commitments
+
+Silverstripe CMS major releases tracks PHP releases. 
```
```suggestion
Silverstripe CMS major releases track PHP releases. 
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+The Silverstripe CMS release cycle is built around the assumptions that:
+- New PHP releases are shipped at the end of November on an annual basis.
+- PHP releases are in full support for two years with 1 year of limited support. 
```
```suggestion
- new PHP releases are shipped at the end of November on an annual basis
- PHP releases are in full support for two years with 1 year of limited support
```
Follows the formatting of other lists

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+- New PHP releases are shipped at the end of November on an annual basis.
+- PHP releases are in full support for two years with 1 year of limited support. 
+
+At launch, a Silverstripe CMS major release support all PHP version in full support. PHP versions in limited support are not supported at launch by new Silverstripe CMS majors.
```
```suggestion
At launch, a Silverstripe CMS major release supports all PHP version in full support. PHP versions in limited support are not supported at launch by new Silverstripe CMS major versions.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+At launch, a Silverstripe CMS major release support all PHP version in full support. PHP versions in limited support are not supported at launch by new Silverstripe CMS majors.
+
+Following the initial release of a Silverstripe CMS major, we will aim to add forward compatibility for the next PHP release. e.g.: Silverstripe CMS 5 at launch will support PHP 8.1 and PHP 8.2. Will aim to add support for an eventual PHP 8.3 release to CMS 5.
```
```suggestion
Following the initial release of a Silverstripe CMS major version, we will aim to add forward compatibility for the next PHP release. e.g.: Silverstripe CMS 5 at launch will support PHP 8.1 and PHP 8.2. Will aim to add support for an eventual PHP 8.3 release to CMS 5.
```
I think we should avoid saying "major" as a noun on its own, and instead say "major release" or "major version" explicitly. Just keeps the language a little more formal which is appropriate for a policy document.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Following the initial release of a Silverstripe CMS major, we will aim to add forward compatibility for the next PHP release. e.g.: Silverstripe CMS 5 at launch will support PHP 8.1 and PHP 8.2. Will aim to add support for an eventual PHP 8.3 release to CMS 5.
+
+We try our best not to drop support for older PHP releases within Silverstripe CMS major release line even if those PHP release reach end-of-life.
```
```suggestion
We try our best not to drop support for older PHP releases within Silverstripe CMS major release line even if those PHP releases reach end-of-life.
```

ChloeSartorelli replied:
Technically this language suggests that we will retain support for EOL php versions even to the detriment of the product, because that would be trying our best. Maybe add something like "but we reserve the right to drop support for such releases if necessary to maintain the security and integrity of the CMS codebase"

xini replied:
correct, as discussed in https://github.com/silverstripe/silverstripe-framework/issues/10345#issuecomment-1171933321. I think it is important that there is one PHP version that you can use during the whole lifecycle of one major SS version.

maxime-rainville replied:
I'll go with "We try our best not to drop support for older PHP releases within Silverstripe CMS major release line even if those PHP release reach end-of-life. Support for end-of-life PHP versions maybe drop if it's necessary to address vulnerabilities or high impact bugs."

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+## Expected Major releases timeline
+
+(Put a pretty Graphic here)
```
Probably link to the roadmap page on .org instead?

maxime-rainville replied:
That page will probably only show current releases ... not future ones.

ChloeSartorelli replied:
Maybe it will make more sense once this graphic exists, then?
I think this PR probably should be in draft still without this graphic though. We can hardly merge it as is.

ChloeSartorelli replied:
Please either add a graphic or remove this placeholder text

maxime-rainville replied:
Pretty picture has been added.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+## Dependency management
+
+At launch new Silverstripe CMS majors aim to have all PHP and JS dependencies on their latest stable version. For dependencies offering a *Long Term Support* version, the latest LTS of that dependency will be preferred.
```
```suggestion
At launch new Silverstripe CMS major releases aim to have all PHP and JavaScript dependencies on their latest stable versions. For dependencies offering a *Long Term Support* version, the latest LTS of that dependency will be preferred.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Support for newer major releases of fixed dependencies will not be retroactively added to Silverstripe CMS major release line, even if the currently used version is end-of-life. The only exception is if the upgrade is necessary to fix a high impact bug or security vulnerability. We’ll only support 1 major release line of each fixed dependency.
+
+Non-fixed dependencies may be upgraded to new major releases or swap out altogether within a Silverstripe CMS release line.
```
```suggestion
Non-fixed dependencies may be upgraded to new major releases or swapped out altogether within a Silverstripe CMS release line.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+The Silverstripe CMS development team will define a list of "fixed dependencies" for each Silverstripe CMS major release line. *Fixed dependencies* will not changed for the lifetime of a Silverstripe CMS major release line.
+
+Support for newer major releases of fixed dependencies will not be retroactively added to Silverstripe CMS major release line, even if the currently used version is end-of-life. The only exception is if the upgrade is necessary to fix a high impact bug or security vulnerability. We’ll only support 1 major release line of each fixed dependency.
```
Supporting multiple major releases of symfony specifically might be a good idea, because packages that some projects will want to use which have symfony dependencies may not be as up-to-date as we are.

maxime-rainville replied:
We'll be running one year behind Symfony. Presuming, people are not in a desperate hurry to jump on the latest major Silverstripe CMS, that should give other vendor a reasonable chance to add support for newer symfony releases.

Symfony is a bit special and there could be a case for treating it a bit like we treat PHP. e.g.: Support for all fully supported majors or add support for new majors as they are release.

Maybe this is something we can revisit once Symfony 7 is released. I would rather start with a stricter support model and loosen it later, rather than start with a loose approach and discover its a major pain in the backside.

ChloeSartorelli replied:
Fair enough, I'm okay revisiting this when we have some real world usage of this policy to reflect on.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+### Symfony support commitments
+
+Symfony is a collection of common PHP libraries Silverstripe CMS uses. At launch, Silverstripe CMS majors will track the latest stable major of Symfony dependencies. A Silverstripe CMS major release line will not add support for later Symfony major release.
```
```suggestion
Symfony is a collection of common PHP libraries Silverstripe CMS uses. At launch, Silverstripe CMS major releases will track the latest stable major release of Symfony dependencies. A Silverstripe CMS major release line will not add support for later Symfony major releases.
```

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Commercially supported modules are reviewed with each new major release. At launch, the list of commercially supported modules for that release line is explicitly defined.
+
+New modules maybe added to the list of commercially for that major release line during its lifespan. We aimed not to withdraw support for modules for the duration of that major release life span.
```
```suggestion
New modules may be added to the list of commercially supported modules for that major release line during its lifespan. We aimed not to withdraw support for modules for the duration of that major release life span.
```

maxime-rainville replied:
I've added a bit in there to provide an example for when we would consider dropping support for module within a major release lifetime.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+New modules maybe added to the list of commercially for that major release line during its lifespan. We aimed not to withdraw support for modules for the duration of that major release life span.
+
+Modules that are supported for one major release line may not be supported for follow up release line.
```
```suggestion
Modules that are supported for one major release line may not be supported for follow up release lines.
```

GuySartorelli reviewed (changes requested):
I've done a grammar/spelling pass for now but I'll have a proper look at this later today from a policy perspective to ensure it covers all the points we've discussed.

GuySartorelli commented:
Where will we be posting information such as the fixed dependencies for a given major release line?

xini commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,135 @@
+
+Following the initial release of a Silverstripe CMS major, we will aim to add forward compatibility for the next PHP release. e.g.: Silverstripe CMS 5 at launch will support PHP 8.1 and PHP 8.2. Will aim to add support for an eventual PHP 8.3 release to CMS 5.
+
+We try our best not to drop support for older PHP releases within Silverstripe CMS major release line even if those PHP release reach end-of-life.
```
I would even go further and align this with the handling of fixed dependencies, because really, that's what PHP is, no?

```suggestion
Within a Silverstripe major release line, we will not drop support for the latest actively maintained PHP release at the time of the initial major Silverstripe release (e.g. PHP 8.2 for Silverstripe 5), even if that PHP release reaches end-of-life. The only exception is if the upgrade is necessary to fix a high impact bug or security vulnerability. 
```

When PHP 8.2 reaches end-of-life in 2025, Silverstripe 5 will only receive bug end security fixes anyway.

maxime-rainville replied:
I'm trying to find a balance between readability, clarity and not painting ourselves in a corner. 

I'll go with:
> Support for end-of-life PHP releases is not dropped within a Silverstripe CMS major release line, unless it's necessary to address vulnerabilities or high impact bugs.

maxime-rainville commented:
I think I've addressed all the feedback and the thing is ready for review.

I've tried to always use the present tense instead of the future tense ... unless I'm describing a specific event (e.g. CMS 5 will be release on ...)

I also tried to stay clear of the "royal we". e.g. "Silverstripe CMS major releases are shipped every two years" instead of "We shipped Silverstripe CMS majors every two years"

GuySartorelli commented:
There's an unanswered question there about where the fixed dependencies and other per-release information will go.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+### Active development
+
+Once a major release is stable, it enters a period of active development. A major release in active development receives:
+- regular minor releases that ships new feature and API in a backward compatible way
```
```suggestion
- regular minor releases that ship new features and API in a backward compatible way
```

maxime-rainville replied:
Done

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+- regular patches for bugs at all impact levels
+- security patches for all vulnerabilities.
+
+A major releases line stays in *active development* for approximatively two years. *Active development* for a major release line ends when the next major release is tagged stable.
```
```suggestion
A major release line stays in *active development* for approximately two years. *Active development* for a major release line ends when the next major release is tagged stable.
```

maxime-rainville replied:
Done

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+
+Silverstripe CMS major releases track PHP releases. 
+
+The Silverstripe CMS release cycle is built around these assumptions.
+
+- New PHP releases are shipped at the end of November on an annual basis.
+- PHP releases are in full support for two years with a year of limited support. 
```
Should this be using the colon format? I.e. are the list items "these assumptions"?
```suggestion
The Silverstripe CMS release cycle is built around these assumptions:

- new PHP releases are shipped at the end of November on an annual basis
- PHP releases are in full support for two years with a year of limited support. 
```

maxime-rainville replied:
done

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+- New PHP releases are shipped at the end of November on an annual basis.
+- PHP releases are in full support for two years with a year of limited support. 
+
+At launch, a Silverstripe CMS major release supports all PHP versions in full support. PHP versions in limited support are not supported at launch by new Silverstripe CMS major release.
```
```suggestion
At launch, a Silverstripe CMS major release supports all PHP versions in full support. PHP versions in limited support are not supported at launch by new Silverstripe CMS major releases.
```
I prefer "versions" for the last word here but "releases" works too, so long as it's plural.

maxime-rainville replied:
done

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+
+At launch, a Silverstripe CMS major release supports all PHP versions in full support. PHP versions in limited support are not supported at launch by new Silverstripe CMS major release.
+
+Following the initial release of a Silverstripe CMS major release, the development team aims to add forward compatibility for the next PHP release. e.g.: Silverstripe CMS 5 at launch will support PHP 8.1 and PHP 8.2. CMS 5 should receive official support for an eventual PHP 8.3 in early 2024.
```
```suggestion
Following the initial release of a Silverstripe CMS major version, the development team aims to add forward compatibility for the next PHP release. e.g.: Silverstripe CMS 5 at launch will support PHP 8.1 and PHP 8.2. CMS 5 should receive official support for an eventual PHP 8.3 in early 2024.
```
"Following the release of a release" sounds pretty bad.
An alternative would be "Following a new Silverstripe CMS Major release,"

maxime-rainville replied:
Switched to "Following the initial launch of a Silverstripe CMS major release"

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+
+Support for end-of-life PHP releases is not dropped within a Silverstripe CMS major release line, unless it's necessary to address vulnerabilities or high impact bugs.
+
+Major releases lines not in active development do not receive official support for later PHP releases.
```
```suggestion
Major release lines not in active development do not receive official support for later PHP releases.
```

maxime-rainville replied:
done

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+The Silverstripe CMS release cycle is built around these assumptions.
+
+- New PHP releases are shipped at the end of November on an annual basis.
+- PHP releases are in full support for two years with a year of limited support. 
```
```suggestion
- PHP releases are in full support for two years with one year of limited support. 
```

maxime-rainville replied:
done

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+
+A list of *fixed dependencies* for each Silverstripe CMS major release line is explicitly defined by the development team. *Fixed dependencies* do not change for the lifetime of a Silverstripe CMS major release line.
+
+Support for newer major releases of fixed dependencies is not added to Silverstripe CMS major release line, even if the currently used version is end-of-life. The only exception is if the upgrade is necessary to fix a high impact bug or security vulnerability. A Silverstripe CMS major release line only support one major version of each fixed dependency.
```
```suggestion
Support for newer major releases of fixed dependencies is not added to Silverstripe CMS major release line, even if the currently used version is end-of-life. The only exception is if the upgrade is necessary to fix a high impact bug or security vulnerability. A Silverstripe CMS major release line only supports one major version of each fixed dependency.
```

maxime-rainville replied:
done

ChloeSartorelli commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -300,13 +300,13 @@ table may be of use:
-| 4.0 - 4.4            | 5.6+        | |
-| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/) |
+| 5.0 +                | 8.1+        | Unreleased            |
+| 4.11 +               | 7.4 - 8.1   | [changelog](/Changelogs/4.11.0#phpeol) |
 | 4.10                 | 7.3+        | [changelog](/Changelogs/4.10.0#phpeol/) |
-| 4.11 +               | 7.4+        | [changelog](/Changelogs/4.11.0#phpeol) |
+| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/)      |
+| 4.0 - 4.4            | 5.6+        |                       |
```
We should either _always_ have a space before the `+` or _never_ have a space before the `+`

maxime-rainville replied:
done

ChloeSartorelli commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -300,13 +300,13 @@ table may be of use:
-| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/) |
+| 5.0 +                | 8.1+        | Unreleased            |
+| 4.11 +               | 7.4 - 8.1   | [changelog](/Changelogs/4.11.0#phpeol) |
 | 4.10                 | 7.3+        | [changelog](/Changelogs/4.10.0#phpeol/) |
```
This isn't accurate - it implies 8.1, 9.x, 10.x etc.
```suggestion
| 4.10                 | 7.3 - 8.0        | [changelog](/Changelogs/4.10.0#php8/) |
```

maxime-rainville replied:
done

ChloeSartorelli commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -300,13 +300,13 @@ table may be of use:
-| 4.0 - 4.4            | 5.6+        | |
-| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/) |
+| 5.0 +                | 8.1+        | Unreleased            |
+| 4.11 +               | 7.4 - 8.1   | [changelog](/Changelogs/4.11.0#phpeol) |
```
```suggestion
| 4.11 +               | 7.4 - 8.1   | [changelog](/Changelogs/4.11.0#php81) |
```
Give the anchor for the relevant item highest on the page. People might scroll down to see other relevant items but are unlikely to scroll up for them.

maxime-rainville replied:
done

ChloeSartorelli commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -300,13 +300,13 @@ table may be of use:
+| 4.11 +               | 7.4 - 8.1   | [changelog](/Changelogs/4.11.0#phpeol) |
 | 4.10                 | 7.3+        | [changelog](/Changelogs/4.10.0#phpeol/) |
-| 4.11 +               | 7.4+        | [changelog](/Changelogs/4.11.0#phpeol) |
+| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/)      |
```
```suggestion
| 4.5 - 4.9            | 7.1 - 7.4        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/)      |
```
No php 8 support

ChloeSartorelli commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -300,13 +300,13 @@ table may be of use:
 | 4.10                 | 7.3+        | [changelog](/Changelogs/4.10.0#phpeol/) |
-| 4.11 +               | 7.4+        | [changelog](/Changelogs/4.11.0#phpeol) |
+| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/)      |
+| 4.0 - 4.4            | 5.6+        |                       |
```
Did 4.0 - 4.4 all support PHP 7.x?

maxime-rainville replied:
Yes ... at launch CMS 4 supported both PHP5.6 and 7. PHP7.2 had just been released.

ChloeSartorelli replied:
It should probably state `5.6 - 7.2` then if those are the supported PHP versions for those minors.

maxime-rainville replied:
From memory, back in those days adding support for newer PHP releases was trivial. I'm pretty sure CMS4.0 would probably have worked OK with PHP7.4.

Given how old those versions are, I don't think it's worth our time to validate exactly in what CMS release we added PHP 7.3 and 7.4 support.

ChloeSartorelli commented on en/05_Contributing/01_Code.md:
```diff
@@ -113,9 +111,16 @@ Other entities might be considered to be included or excluded from the public AP
 Any updates to third party dependencies in composer.json should aim to target the default branch for a minor release if possible. Targeting a patch release branch is acceptable if updating dependencies is required to fix a bug.
 
+API from third party dependencies may indirectly be incorporated into Silverstripe CMS's API if:
+- they are define as a parameter type for a supported method
+- they are define as a return type for a supported method
```
```suggestion
- they are defined as a parameter type for a supported method
- they are defined as a return type for a supported method
```

maxime-rainville replied:
done

ChloeSartorelli commented on en/05_Contributing/01_Code.md:
```diff
@@ -113,9 +111,16 @@ Other entities might be considered to be included or excluded from the public AP
+API from third party dependencies may indirectly be incorporated into Silverstripe CMS's API if:
+- they are define as a parameter type for a supported method
+- they are define as a return type for a supported method
+- they are extended by a Silverstripe CMS class.
```
What about if they are provided as an explicitly injectable service?

maxime-rainville replied:
"explicitly injectable service" didn't get brought up in the RFC. It probably should be discussed in a follow up discussion.

ChloeSartorelli replied:
https://github.com/silverstripe/silverstripe-framework/issues/10345#issuecomment-1155769208
> I think that it should include considerations around the fact that a developer could access (say) `protected` methods and properties through `Injector`.

It was briefly mentioned, though using different wording and didn't really get any traction as a talking point. I'm okay with it being discussed in a follow-up discussion so long as it does _actually_ get discussed, because IMO if we're setting up an `Injectable` service people should be able to rely on that service existing and being of a consistent type.

maxime-rainville replied:
I can't remember the last time a PR came up where this would have been an issue. It's mostly an academic point. Resolving this ambiguity is a low priority from my end. 

If you want to create a follow up PR to clarify when an Injectable object should be part of our API, I'm happy to have a look at it.

ChloeSartorelli replied:
A follow-up PR would likely be pointless at this stage - I already made a recommendation (albeit in the format of a question) and it was effectively rejected at this stage, with the suggestion of a follow-up discussion. It sounds like discussion is the next step. That needn't hold up this PR, so I'm fine with that discussion happening separately from this PR after this is merged.

ChloeSartorelli commented on en/05_Contributing/05_Making_A_SilverStripe_Core_Release.md:
```diff
@@ -91,7 +91,7 @@ SS_BASE_URL="http://localhost/"
 
 You will also need to be assigned the following permissions. Contact one of the SilverStripe staff from
-the [core committers](core_committers), who will assist with setting up your credentials.
+the [core committers](/project_governance/core_committers), who will assist with setting up your credentials.
```
There's another one of these at the bottom of the page to update.

maxime-rainville replied:
Fixed that link and a few minor unrelated links.

xini commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -300,13 +300,13 @@ table may be of use:
-| 3.7                  | 5.3 - 7.4   | [changelog](https://docs.silverstripe.org/en/3/changelogs/3.7.4/) |
-| 4.0 - 4.4            | 5.6+        | |
-| 4.5 - 4.9            | 7.1+        | [blog post](https://www.silverstripe.org/blog/our-plan-for-ending-php-5-6-support-in-silverstripe-4/) |
+| 5.0 +                | 8.1+        | Unreleased            |
```
```suggestion
| 5.0 +                | 8.1 - 8.2   | Unreleased            |
```

maxime-rainville replied:
Fixed

maxime-rainville commented:
> Where will we be posting information such as the fixed dependencies for a given major release line?

I suspect we'll have a dedicated page. I don't think we need to figure this out at this time.

ChloeSartorelli commented on en/06_Project_Governance/05_Major_release_policy.md:
```diff
@@ -0,0 +1,143 @@
+
+This timeline is provided as an indication only.
+
+![Timeline of planned Silverstripe CMS major releases](../../_images/major_release_timeline.jpg)
```
Now that there is a graphic I am even more certain that this belongs on the roadmap page rather than here, since it is talking about _specific_ releases (and will therefore eventually be outdated), where this policy is intended to be release-agnostic - it doesn't care about and _specific_ releases.

[Your reasoning last time](https://github.com/silverstripe/developer-docs/pull/54#discussion_r939484942) I mentioned the roadmap page for this was:
> That page will probably only show current releases ... not future ones.

But a roadmap is really meant to be about the future, so I think that page is a perfect fit for a graphic that explicitly talks about the intended release schedule of specific future releases.

I don't think we need a graphic for this policy, but if we do have one I think it should not refer to any specific releases. This whole section in fact probably just doesn't really belong here - it should be on the roadmap page where people will be explicitly looking for release timeline info.

ChloeSartorelli replied:
All that said, I don't feel super strongly about this - so if you want to keep the graphic here I'm okay with it in the interest of getting this over the line.

maxime-rainville replied:
The main value of the graphic is that it visually conveys what releases will happen one and the overlap between releases. 

Happy to move the graphic to an eventual roadmap page once SSORG is rebuilt. In the short term, I think it provides value.

ChloeSartorelli commented on en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -299,14 +299,14 @@ Silverstripe CMS's PHP support has changed over time and if you are looking to u
+| ------------------------ | ----------- | ---------------- |
+| 5.0 +                    | 8.1 - 8.2   | Unreleased       |
+| 4.11 +                   | 7.4 - 8.1   | [changelog](/Changelogs/4.11.0#php81) |
+| 4.10                     | 7.3 - 8.0   | [changelog](/Changelogs/4.10.0#phpeol/) |
```
```suggestion
| 4.10                     | 7.3 - 8.0   | [changelog](/Changelogs/4.10.0#phpeol) |
```
The anchor won't work with a slash at the end.

GuySartorelli reviewed (changes requested):
One very small change (removing the slash from an anchor link) to be done - the rest are effectively commentary that you can action/respond to/etc as you please.
I'll be happy to merge when the slash is removed from that one link.
